internal/slack/translator.go

```diff
@@ -0,0 +1,233 @@
+		return "", err
+	}
+
+	attachmentDirName := fmt.Sprintf("%s/attachments", workdir)
+	archiveWithFilesName, err := st.addFilesToSlackArchive(
+		logger,
+		workdir,
+		attachmentDirName,
+		inputArchiveName,
+	)
+
+	mbifName := fmt.Sprintf("%s/%s_MBIF.jsonl", workdir, translation.InstallationID)
+	logger.Infof("Transforming Slack archive for Translation %s to MBIF", translation.ID)
+	err = TransformSlack(
+		archiveWithFilesName,
+		mbifName,
+		translation.Team,
+		attachmentDirName,
+	)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to transform Slack archive to MBIF")
+	}
+
+	outputZip, err := st.createOutputZipfile(logger, attachmentDirName, mbifName, translation.ID)
+	if err != nil {
+		return "", err
+	}
+
+	err = st.uploadTransformedZip(outputZip, st.bucket)
+	if err != nil {
+		return "", err
+	}
+
+	outputNameSplitPath := strings.Split(outputZip, "/")
+	outputShortName := outputNameSplitPath[len(outputNameSplitPath)-1]
+	logger.Infof("Finished translation %s", translation.ID)
+
+	return outputShortName, nil
+}
+
+// fetchSlackArchive is responsible for downloading the input archive
+// from S3 and writing it out to workdir, which is assumed to be of
+// sufficient capacity for the archive
+func (st *SlackTranslator) fetchSlackArchive(logger logrus.FieldLogger, workdir, resource string) (string, error) {
+
+	sess := session.Must(session.NewSession())
+
+	downloader := s3manager.NewDownloader(sess)
+
+	inputArchive, err := os.Create(workdir + "/input.zip")
+	if err != nil {
+		return "", errors.Wrap(err, "failed to open temp file to download input archive to")
+	}
+
+	nBytes, err := downloader.Download(inputArchive,
+		&s3.GetObjectInput{
+			Bucket: &st.bucket,
+			Key:    &resource,
+		})
+
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to download %s from bucket %s", resource, st.bucket)
+	}
+
+	err = inputArchive.Close()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to close temporary file after writing incoming archive to it")
+	}
+
+	logger.Debugf("Successfully downloaded %d bytes from bucket %s key %s",
+		nBytes, st.bucket, resource)
+
+	return inputArchive.Name(), nil
+}
+
+// addFilesToSlackArchive prepares the input and fetches attached
+// files, writing the output to workdir and removing the input archive
+// when complete
+func (st *SlackTranslator) addFilesToSlackArchive(logger logrus.FieldLogger, workdir, attachmentDirName, inputArchiveName string) (string, error) {
+
+	logger.Infof("Downloading attached files to %s", attachmentDirName)
+
+	withFiles, err := os.Create(workdir + "/inputWithFiles.zip")
+	if err != nil {
+		return "", errors.Wrap(err, "failed to open temp file to convert input archive to")
+	}
+
+	err = FetchAttachedFiles(inputArchiveName, withFiles.Name())
+	if err != nil {
+		return "", errors.Wrap(err, "failed to fetch attached files")
+	}
+
+	err = os.Remove(inputArchiveName)
+	if err != nil {
+		logger.Errorf("failed to remove file %s", inputArchiveName)
+	}
```

Comment:
I wonder if it would not be better to defer archive removal at the top level?

Reply:
Hmm, I know I specifically chose not to use a defer here but I can't remember why anymore :stuck_out_tongue: . I think I did this either to ease some pain related to development/testing or because I wanted to capture that error message? But I can't remember what pain I was alleviating and there's no reason I can't capture the error message in a deferred inline func so I'll do that..